rating/peersRatingHandler.go

```diff
@@ -129,10 +183,88 @@ func (prh *peersRatingHandler) getOldRating(pid core.PeerID) (int32, bool) {
 	return defaultRating, found
 }
 
-func (prh *peersRatingHandler) updateRatingIfNeeded(pid core.PeerID, updateFactor int32) {
-	defer prh.updateMetrics()
+func (prh *peersRatingHandler) markInactivePIDsForRemoval(receivedPIDs map[string]struct{}) {
+	removalTimestamp := prh.getTimeHandler().Add(prh.timeWaitingForReconnection).Unix()
+	topRatedPIDs := prh.topRatedCache.Keys()
+	for _, pid := range topRatedPIDs {
+		_, isPIDStillActive := receivedPIDs[string(pid)]
+		alreadyMarked := prh.markedForRemovalCache.Has(pid)
+		if !isPIDStillActive && !alreadyMarked {
+			prh.markedForRemovalCache.Put(pid, removalTimestamp, int64Size)
+		}
+	}
+	badRatedPIDs := prh.badRatedCache.Keys()
+	for _, pid := range badRatedPIDs {
+		_, isPIDStillActive := receivedPIDs[string(pid)]
+		alreadyMarked := prh.markedForRemovalCache.Has(pid)
+		if !isPIDStillActive && !alreadyMarked {
+			prh.markedForRemovalCache.Put(pid, removalTimestamp, int64Size)
+		}
+	}
+}
+
+func (prh *peersRatingHandler) sweepCachersLoop(ctx context.Context) {
+	timer := time.NewTimer(prh.timeBetweenCachersSweep)
+	defer timer.Stop()
 
-	oldRating, found := prh.getOldRating(pid)
+	for {
+		timer.Reset(prh.timeBetweenCachersSweep)
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+		}
+		prh.sweepCachers()
+	}
+}
+
+func (prh *peersRatingHandler) sweepCachers() {
+	prh.mut.Lock()
+	defer prh.mut.Unlock()
+
+	pids := prh.markedForRemovalCache.Keys()
+	for _, pidBytes := range pids {
+		value, exists := prh.markedForRemovalCache.Get(pidBytes)
+		if !exists {
+			continue
+		}
+
+		removalTimestamp, ok := value.(int64)
+		if !ok {
+			log.Error("could not parse data to int64")
+			continue
+		}
+
+		if removalTimestamp <= prh.getTimeHandler().Unix() {
+			prh.removePIDFromCachers(pidBytes)
+		}
+	}
+}
+
+func (prh *peersRatingHandler) removePIDFromCachers(pid []byte) {
```

Comment:
having a second thought here, I think we can leak memory if the markedForRemovalCache is improperly sized and it starts evicting items. What do you think: would it be better to switch to an internal map?

Reply:
indeed, it would be very hard to have this cache perfectly aligned with the others.. switched to map